### PR TITLE
Leave-one-out cross validation

### PR DESCRIPTION
**Pull request by fonnesbeck**
Implements LOO cross-validation using Pareto-smoothed cross-validation, following Vehtari et al (2015).

### CODE REVIEW

**Comment by fonnesbeck** [comment]
I added a notebook with a model comparison example. The DIC values look a little suspicious. I wonder if we are calculating it correctly.

File: pymc3/stats.py [hunk 1]
```diff
@@ -7,11 +7,12 @@
 import warnings
 from .model import modelcontext
 
+import scipy.stats.distributions  as sp
```
Comment (twiecki) [inline pymc3/stats.py]: usually just scipy gets imported as `sp`. Instead maybe just import `pareto` directly?

File: pymc3/stats.py [hunk 2]
```diff
@@ -96,12 +97,18 @@ def dic(trace, model=None):
...
+    Calculate the elementwise log-posterior for the sampled trace.
+    '''
+    return np.hstack([obs.logp_elemwise(pt) for pt in trace] for obs in model.observed_RVs)
 
 def waic(trace, model=None):
     """
     Calculate the widely available information criterion of the samples in trace from model.
     Read more theory here - in a paper by some of the leading authorities on Model Selection - http://bit.ly/1W2YJ7c
-    """
+    """
```
Comment (twiecki) [inline pymc3/stats.py]: Stray tab.

File: pymc3/stats.py [hunk 3]
```diff
@@ -112,17 +119,60 @@ def waic(trace, model=None):
...
+    model fit, following Vehtari et al. (2015). Cross-validation is computed using
+    Pareto-smoothed importance sampling (PSIS).
+    
+    Returns log pointwise predictive density calculated via approximated LOO cross-validation.
+    """
+    model = modelcontext(model)
+    
+    log_py = log_post_trace(trace, model)
+    
+    q = int(len(log_py)*0.8)
```
Comment (twiecki) [inline pymc3/stats.py]: Is this used anywhere? below you compute `q80`.

File: pymc3/stats.py [hunk 4]
```diff
@@ -112,17 +119,60 @@ def waic(trace, model=None):
...
+    
+    # Importance ratios
+    r = 1./np.exp(log_py)
+    r_sorted = np.sort(r, axis=0)
+
+    # Extract largest 20% of importance ratios and fit generalized Pareto to each 
+    # (returns tuple with shape, location, scale)
+    q80 = int(len(log_py)*0.8)
+    pareto_fit = np.apply_along_axis(lambda x: sp.pareto.fit(x, floc=0), 0, r_sorted[q80:])
```
Comment (twiecki) [inline pymc3/stats.py]: instead, couldn't you do `r[r > np.percentile(r, 80)]` ?

File: pymc3/stats.py [hunk 5]
```diff
@@ -112,17 +119,60 @@ def waic(trace, model=None):
...
+    z = (np.arange(M)+0.5)/M
+    expvals = map(lambda x: sp.pareto.ppf(z, x[0], scale=x[2]), pareto_fit.T)
+    
+    # Replace importance ratios with order statistics of fitted Pareto
+    r_sorted[q80:] = np.vstack(expvals).T
+    
+    # Truncate weights to guarantee finite variance
+    w = np.minimum(r_sorted, r_sorted.mean(axis=0) * S**0.75)
+    
+    loo_lppd =  np.sum(np.log(np.sum(w * np.exp(log_py), axis=0) / np.sum(w, axis=0)))
```
Comment (twiecki) [inline pymc3/stats.py]: stray space

File: pymc3/stats.py (hunk 4 above)
Comment (fonnesbeck) [inline pymc3/stats.py]: That doesn't work. It will return a 1-d array. That did help me find a bug though!

**Comment by fonnesbeck** [comment]
This should be good now. I had forgotten to unsort the ratios before using them as weights.

**Comment by twiecki** [comment]
Looks great!

**Pull request merged**